tests/cpp/test_multidevice_transformer.cpp

```diff
@@ -535,6 +798,111 @@ TEST_P(DistributedTransformerTest, MLP_Backward) {
   validate(expected_outputs, outputs);
 }
 
+TEST_P(DistributedTransformerTest, MHA_Backward) {
+  auto dtype = GetParam();
+  at::ScalarType at_dtype = data_type_to_aten(dtype);
+  auto fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+  const auto mesh = DeviceMesh::createForNumDevices(D);
+
+  TensorView* tvx = makeContigConcreteTensor({B * S, E}, dtype);
+  TensorView* tvw0 = makeContigConcreteTensor({D, E, 3 * E / D}, dtype);
+  TensorView* tvb0 = makeContigConcreteTensor({D, 3 * E / D}, dtype);
+  TensorView* tvw1 = makeContigConcreteTensor({D, E / D, E}, dtype);
+  TensorView* tvgrad = makeContigConcreteTensor({B * S, E}, DataType::Float);
+  TensorView* tvmask = makeContigConcreteTensor({B * S, E}, DataType::Bool);
+  TensorView* tvsdpa_out =
+      makeContigConcreteTensor({D, B, H / D, S, E / H}, dtype);
+  TensorView* tvsdpa_log_sumexp =
+      makeContigConcreteTensor({D, B, H / D, S}, DataType::Float);
+  TensorView* tvsdpa_seed = makeSymbolicTensor({}, DataType::Int);
+  TensorView* tvspda_offset = makeSymbolicTensor({}, DataType::Int);
+
+  fusion->addInput(tvx);
+  fusion->addInput(tvw0);
+  fusion->addInput(tvb0);
+  fusion->addInput(tvw1);
+  fusion->addInput(tvgrad);
+  fusion->addInput(tvmask);
+  fusion->addInput(tvsdpa_out);
+  fusion->addInput(tvsdpa_log_sumexp);
+  fusion->addInput(tvsdpa_seed);
+  fusion->addInput(tvspda_offset);
+
+  auto tvouts = mha_backwards(
+      tvx,
+      tvw0,
+      tvb0,
+      tvw1,
+      tvmask,
+      tvsdpa_out,
+      tvsdpa_log_sumexp,
+      tvsdpa_seed,
+      tvspda_offset,
+      tvgrad,
+      mesh,
+      dtype);
+
+  for (auto tv : tvouts) {
+    fusion->addOutput(tv);
+  }
+
+  shardBetween(
+      {tvw1, tvw0, tvb0},
+      {tvouts[1],
+       tvouts[2],
+       tvouts[3],
+       tvouts[4],
+       tvouts[5],
+       tvouts[6],
+       tvouts[7]},
+      tvw0);
+  shardBetween({tvx, tvmask, tvgrad}, {tvouts[0], tvouts[8]}, tvx);
```

Comment:
Can you code-comment on why it's sharded this way? Some illustrative ASCII art may help.

Reply:
Comment added. The propagation is relatively simple, just have to cover all the outputs so that's why the list is long.

Reply:
+ reduced the set in `to` since some were not leaves of the DAG and were covered by other TVs.